### DISCUSSION TITLE
pocodex: the unofficial PocketBase plugin and code exchange

### DISCUSSION THREAD

**Discussion by benallfree** [pocketbase/pocketbase#5542] [Show and tell]
Introducing **pocodex**, the POcketbase CODe EXchange. (*Please note, Gani wanted me to emphasize that this is an unofficial project.*)

https://pocodex.dev

This is the PocketBase JS plugin system talked about a lot (#1196  #3679 #5415 [PocketHost Discord](https://discord.com/channels/1128192380500193370/1215133123294011463/1215165389835866135) and more). pocodex defines a plugin ecosystem and architecture for extending PocketBase using JS hooks like many popular CMS have.

## What plugins can be made?

Some plugins already exist. Small ones like [pocketbase-otp](https://npmjs.com/pocketbase-otp) for One-Time Passwords, and more ambitious plugins like [PocketPages](https://nodejs.com/pocketpages), which is an entire Server-side JS pages rendering framework. You can install and uninstall these at will. 

You can also create plugins for end-users:

* admins
* blogs
* storefronts
* personal portfolios
* apps
* podcasts

(plugins I personally plan to make include a blog and storefront)

## Architecture

pocodex is both a plugin manager and a standard for writing plugins. It defines a technical architecture that addresses key concerns needed to make plugins work well with the PocketBase JSVM and migration system.

* installing/uninstalling/updating plugins
* a build system to bundle code and packages
* managing up and down migrations
* persistent key/value storage (to avoid unnecessary full migrations)
* version compatibility

## Quickstart

```bash
bunx pocodex init --trust
pocketbase pocodex install <plugin> --dir=pb_data
pocketbase serve --dir=pb_data
```

## Call for plugin authors

What I need most now is community feedback and a few early adopters to create plugins and list them in the pocodex directory. I'll continue contributing plugins and developing the core features.

**Comment by Leftium** [comment]
Two plugins I would be interested in developing:
1. A proxy API that allows an SPA to fetch() from the client despite CORS restrictions.
2. Guest/anonymous accounts. Probably implemented as described here: https://github.com/pocketbase/pocketbase/discussions/3035

I planned to develop these as hooks anyways; if it's not too much extra effort I would be happy to develop them as pocodex plugins.

**Reply by benallfree** [reply]
Jump over to the pocodex repo to discuss 🤙
